coord/src/rmr/rmr.c

```diff
@@ -660,38 +557,26 @@ bool MR_ManuallyTriggerNextIfNeeded(MRIterator *it, size_t channelThreshold) {
   return channelSize > 0;
 }
 
-MRIterator *MR_Iterate(MRCommandGenerator cg, MRIteratorCallback cb) {
+MRIterator *MR_Iterate(const MRCommand *cmd, MRIteratorCallback cb) {
 
-  MRIterator *ret = rm_malloc(sizeof(*ret));
-  size_t len = cg.Len(cg.ctx);
+  MRIterator *ret = rm_new(MRIterator);
+  // Initial initialization of the iterator.
+  // The rest of the initialization is done in the iterator start callback.
+  // We set `pending` and `inProcess` to 1 so we won't get the impression that we are done
+  // before the first command is sent. This is also technically correct since we know that we have
+  // at least ourselves to wait for.
   *ret = (MRIterator){
-      .ctx =
-          {
-              .cluster = cluster_g,
-              .chan = MR_NewChannel(0),
-              .cb = cb,
-              .pending = 0,
-              .timedOut = 0,
-          },
-      .cbxs = rm_calloc(len, sizeof(MRIteratorCallbackCtx)),
-      .len = len,
+    .ctx = {
+      .chan = MR_NewChannel(),
+      .cb = cb,
+      .pending = 1,
+      .inProcess = 1,
+      .timedOut = 0,
+    },
+    .cbxs = rm_new(MRIteratorCallbackCtx),
   };
-
-  for (size_t i = 0; i < len; i++) {
-    ret->cbxs[i].ic = &ret->ctx;
-    if (!cg.Next(cg.ctx, &(ret->cbxs[i].cmd))) {
-      ret->len = i;
-      break;
-    }
-  }
-
-  // Could not create command, probably invalid cluster
-  if (ret->len == 0) {
-    MRIterator_Free(ret);
-    return NULL;
-  }
-  ret->ctx.pending = ret->len;
-  ret->ctx.inProcess = ret->len; // Initially all commands are in process
+  // Temporary copy of the command.
+  ret->cbxs->cmd = *cmd;
```

Comment:
Do we know it will not cause use after free accidentally? If we will free the cmd and then use it?

Reply:
we won't free the command or the iterator before there are no pending commands.
I can still change it so on this line we pass the ownership over it, and free it from the UV thread on the `iterStartCB`

Reply:
If we can pass ownership I think its better.